### PR TITLE
Add local storage save logic for search based creation UI

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/21098

This PR adds autosave logic for search-based insight and code stats insight creation UI. 

if the user filled in some fields and left the page. When returning to the form page, it will see the last value of the filled fields.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]

**Review by tjkandala — APPROVED** [review]

File: client/web/src/insights/pages/creation/search-insight/components/search-insight-creation-content/SearchInsightCreationContent.tsx [hunk 1]
```diff
@@ -91,6 +95,18 @@ export const SearchInsightCreationContent: React.FunctionComponent<SearchInsight
...
+    const handleFormReset = (): void => {
+        // TODO [VK] Change useForm API in order to implement form.reset method.
+        title.input.onChange('')
+        repositories.input.onChange('')
+        // Focus first element of the form
+        repositories.input.ref.current?.focus()
+        visibility.input.onChange('personal')
+        series.input.onChange([createDefaultEditSeries({ edit: true })])
+        stepValue.input.onChange('2')
+        step.input.onChange('months')
```
Comment (valerybugakov) [inline client/web/src/insights/pages/creation/search-insight/components/search-insight-creation-content/SearchInsightCreationContent.tsx:107]: Does it make sense to use `initialValue` here?

File: client/web/src/insights/pages/creation/lang-stats/components/lang-stats-insight-creation-content/LangStatsInsightCreationContent.tsx [hunk 2]
```diff
@@ -78,6 +83,18 @@ export const LangStatsInsightCreationContent: React.FunctionComponent<LangStatsI
         repository.meta.validState === 'VALID' ||
         (repository.meta.validState === 'CHECKING' && threshold.meta.validState === 'VALID')
 
+    const handleFormReset = (): void => {
+        // TODO [VK] Change useForm API in order to implement form.reset method.
```
Comment (valerybugakov) [inline client/web/src/insights/pages/creation/lang-stats/components/lang-stats-insight-creation-content/LangStatsInsightCreationContent.tsx:87]: 👍

**Review by valerybugakov — APPROVED** [review]

File: client/web/src/insights/components/form/hooks/useField.ts [hunk 3]
```diff
@@ -130,7 +130,15 @@ export function useField<FormValues, FieldValueKey extends keyof FormAPI<FormVal
...
         if (async) {
-            startAsyncValidation({ value: state.value, validity })
+            // Due the call of start async validation in useLayoutEffect we have to
+            // schedule the async validation event in the next tick to be able run
+            // observable pipeline validation since useAsyncValidation hook use
+            // useObservable hook internally which calls '.subscribe' in useEffect.
+            // In order to catch start async validation event we need until next
+            // tick.
+            requestAnimationFrame(() => {
+                startAsyncValidation({ value: state.value, validity })
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/insights/components/form/hooks/useField.ts:140]: Nice tip. I created an issue for it, thanks. https://github.com/sourcegraph/sourcegraph/issues/21909

File: client/web/src/insights/pages/creation/search-insight/components/search-insight-creation-content/SearchInsightCreationContent.tsx (hunk 1 above)
Comment (vovakulikov) [reply client/web/src/insights/pages/creation/search-insight/components/search-insight-creation-content/SearchInsightCreationContent.tsx:107]: You're right. I was thinking about this too. But in the case when we're using this component on the Edit page initial value might be a bit different from our standard initial value for the creation page. 
I will refine these handlers in a separate ticket where I change `useForm` API for the reset method.

Related ticket https://github.com/sourcegraph/sourcegraph/issues/21911

**Pull request merged**